tensorrt_llm/_torch/auto_deploy/transformations/library/rope.py

```diff
@@ -1,115 +1,461 @@
+                    "Unable to infer layout of q node. Skip layout matching"
+                )
+
+            need_transpose = expected_layout.lower() != current_layout
+
+            if need_transpose:
+                ad_logger.debug(
+                    f"Inferred RoPE input layout: '{current_layout}']"
+                    f"Mapping layout to '{expected_layout}']"
+                )
+                with graph.inserting_before(node):
+                    q_for_op = graph.call_function(torch.ops.aten.transpose, args=(q_node, 1, 2))
+                    k_for_op = graph.call_function(torch.ops.aten.transpose, args=(k_node, 1, 2))
+                    q_for_op_contig = graph.call_function(
+                        torch.ops.aten.contiguous, args=(q_for_op,)
+                    )
+                    k_for_op_contig = graph.call_function(
+                        torch.ops.aten.contiguous, args=(k_for_op,)
+                    )
+
+                q_for_op_contig.meta["val"] = q_node.meta["val"].transpose(1, 2)
+                k_for_op_contig.meta["val"] = k_node.meta["val"].transpose(1, 2)
+
+                if is_op(node, torch.ops.rope.torch_apply_rope_with_complex_freqs):
+                    new_args = (
+                        q_for_op_contig,
+                        k_for_op_contig,
+                        freqs_node,
+                        2 if expected_layout.lower() == "bsnd" else 1,
+                    )  # unsqueeze_dim updated
+                else:
+                    new_args = (
+                        q_for_op_contig,
+                        k_for_op_contig,
+                        cos_node,
+                        sin_node,
+                        2 if expected_layout.lower() == "bsnd" else 1,
+                    )  # unsqueeze_dim updated
+                node.args = new_args
+
+                # retrieve q and k output node from node
+                old_q, old_k = extract_output_tuple(node, 2)
+                if old_q is None or old_k is None:
+                    ad_logger.warning(
+                        f"Failed to extract all two outputs from the explicit op, \
+                            get {old_q}, {old_k}, fail to replace {node} with flashinfer rope"
+                    )
+                    return
+                with graph.inserting_after(old_q):
+                    new_q = graph.call_function(torch.ops.aten.transpose, args=(old_q, 1, 2))
+                with graph.inserting_after(old_k):
+                    new_k = graph.call_function(torch.ops.aten.transpose, args=(old_k, 1, 2))
+
+                # Preserve fake tensor in meta["val"] for the transposed inputs
+                old_q.meta["val"] = old_q.meta["val"].transpose(1, 2)
+                new_q.meta["val"] = old_q.meta["val"].transpose(1, 2)
+                old_k.meta["val"] = old_k.meta["val"].transpose(1, 2)
+                new_k.meta["val"] = old_k.meta["val"].transpose(1, 2)
```

Comment:
```suggestion
                new_q.meta["val"] = old_q.meta["val"].transpose(1, 2)
                new_k.meta["val"] = old_k.meta["val"].transpose(1, 2)
```

why update old_q metadata and then update new_q metadata with the updated old_q metadata? This seems wrong

Reply:
When we change the layout of the input to the custom op, the output of it also changed, so I put `old_q.meta["val"] = old_q.meta["val"].transpose(1, 2)`.  `new_q` is `old_q.transpose(1,2)`, so it is actually of the same shape as previous output of the custom op. 
I changed the code a bit, hope it's less confusing.

Reply:
gotcha. makes sense now